README.rst
==============================
 nti.plone.namedfile 
==============================

.. image:: https://travis-ci.org/NextThought/nti.plone.namedfile .svg?branch=master
    :target: https://travis-ci.org/NextThought/nti.plone.namedfile 

.. image:: https://coveralls.io/repos/github/NextThought/nti.plone.namedfile /badge.svg?branch=master
    :target: https://coveralls.io/github/NextThought/nti.plone.namedfile ?branch=master


Introduction
============

This package contains fields and wrapper objects for storing:

* A file with a filename
* An image with a filename

Blob-based and non-blob-based types are provided. 

Source Code
===========

 Note: This packages is licensed under a *Apache license*. 
 Please do not add dependencies on GPL code!

Sources are at the `Plone code repository hosted at Github <https://github.com/NextThought/nti.plone.namedfile>`_.


CHANGES.rst
=========
 Changes
=========

5.0.0 (unreleased)
------------------

New features:

- Add support for PyPy
- Add support for Python 3
- Remove ``unicode_literals``.
- Refactor testing layer

Incompatibilities:

- Remove the entire Plone Stack dependencies and associated code
- Remove zope schema definitions
- Remove marshaller classes
- Remove scaling classes


4.2.4 (unreleased)
------------------

Bug fixes:

- remove mention of "retina" (https://github.com/plone/Products.CMFPlone/issues/2123)
  [tkimnguyen]


4.2.3 (2017-09-08)
------------------

Bug fixes:

- Fix bug #56 where ``srcset`` generation failed on no given width or height if there was no sclae given.
  https://github.com/plone/plone.namedfile/pull/56
  [jensens]


4.2.2 (2017-07-03)
------------------

Bug fixes:

- Dont't break DefaultImageScalingFactory, if for any reason the fieldname isn't available on the context.
  [thet]

- Different caching keys for different domains
  [mamico]


4.2.1 (2017-05-30)
------------------

Bug fixes:

- Fix #46, when ``process_png``, ``process_jpeg`` and ``process_tiff`` could fail with a ``width referenced before assignment`` error.
  [thet]

- Fix contentType attribute should be str type, what leads to validation errors (fixes `#38`_).
  [rodfersou]

- Fix bug on Image rotation if ImageIFD.XResolution or ImageIFD.YResolution are not set.
  [loechel]

- Fix: Do not log failing PIL image regognition as error, but as warning.
  [jensens]

- Fix: compatibility for Plone 4 re-added.
  [loechel]


4.2.0 (2017-03-26)
------------------

New features:

- Add retina image scales using srcset attribute.
  [didrix]


4.1.2 (2017-02-12)
------------------

Bug fixes:

- BrowserViews have no Acquisition.
  [pbauer]


4.1.1 (2017-01-20)
------------------

New features:

- Add automatic image rotation based on EXIF data for all images.
  Based on piexif library and ideas of maartenkling and ezvirtual.
  Choosen piexif as it allow read and write of exif data for future enhancements.
  http://piexif.readthedocs.org/en/latest/
  For Orientation examples and description see http://www.daveperrett.com/articles/2012/07/28/exif-orientation-handling-is-a-ghetto/ test data https://github.com/recurser/exif-orientation-examples
  Additional Test Images with different MIME-Types (JPEG and TIFF) and possible problems: https://github.com/ianare/exif-samples.git
  [loechel]

- Support SVG images
  [tomgross]


Bug fixes:

- Added handler for Tiff Images in getImageInfo.
  [loechel]

- Restructured packages.
  Moved image meta data detection in an own subfolder
  [loechel]


4.1 (2016-09-14)
----------------

New features:

- Add Pdata storage
  [vangheem]


4.0 (2016-08-12)
----------------

Incompatibilities:

- Targets Plone 5.1 only, coredev 5.0 and 4.3 are on 3.0.x branch [jensens]:

  - ``plone.supermodel``, ``plone.scale`` and ``plone.schemaeditor`` are now hard depedencies.
    The extras  in setup.py are kept for bbb reasons, but are empty.
    Conditional code is now no longer conditional.
    This simplifies the code a lot.

  - ``zope.app.file`` is no longer hard dependency.
    If it is there, its FileChunk implementation is still checked for, otherwise not.


New:

- uses adapter as factory for scales as in plone.scale>=1.5
  [jensens]

Fixes:

- Several tests were failing on Windows 10 due to binary files being opened in text mode. Fixed.
  [smcmahon]

- Prevent attempt to create a filestream_iterator from a temporary file associated with an
  uncommited blob.
  Fixes an error on Windows 10 "WindowsError 32" by attempting to delete or access a file in use
  by another process.
  [smcmahon]

- Fix tests to work with latest plone.scale changes, where gif images are no longer converted to jpeg.
  [thet]

- Fixed test setup to use layers properly.
  [jensens]

- Fixed test isolation problem in ``test_blobfile.py``.
  [jensens]

- Fix warning on testing.zcml missing an i18n:domain.
  [gforcada]

- Fix some code analysis warnings.
  [gforcada]

3.0.8 (2016-02-26)
------------------

Fixes:

- PEP 8, UTF-8 headers, implements/adapts to decorators, doctest formating.
  [thet, jensens]

- Workarround for method getImageSize.
  Prevent returning (-1, -1) as the size of the image.
  [andreesg]


3.0.7 (2016-02-12)
------------------

Fixes:

- Make plone.protect a soft dependency. This allows to use this package in
  setups without the Plone stack. Fixes plone/Products.CMFPlone#1311
  [thet]

3.0.6 (2016-01-08)
------------------

Fixes:

- Stabilised tests.  [gotcha]


3.0.5 (2015-11-26)
------------------

New:

- Added webdav support to image scales.
  https://github.com/plone/Products.CMFPlone/issues/1251
  [maurits]


3.0.4 (2015-10-28)
------------------

Fixes:

- No longer rely on deprecated ``bobobase_modification_time`` from
  ``Persistence.Persistent``.
  [thet]


3.0.3 (2015-08-14)
------------------

- Don't fail, when accessing the ``tag`` method of the ``@@images`` view, if
  ``scale`` returns ``None``.
  [thet]


3.0.2 (2015-03-13)
------------------

- Cache image scales using the plone.stableResource ruleset when they are
  accessed via UID-based URLs. (Requires plone.app.imaging >= 1.1.0)
  [davisagli]


3.0.1 (2014-10-23)
------------------

- Fixed inserting filename in Content-Disposition header.
  [kroman0]

- Respect field level security in download views also for primary fields.
  [jensens]

- Internationalize field factory label.
  [thomasdesvenain]


3.0.0 (2014-04-13)
------------------

- Disable CSRF protection when creating a scale so we can write to the database
  [vangheem]


2.0.5 (2014-02-19)
------------------

- Ensure zope.app.file.file module alias is created before its use in
  file package.
  [thomasdesvenain]


2.0.4 (2014-01-27)
------------------

- Disable CSRF protection when creating a scale so we can write to the database
  [vangheem]

- Validate image field : check if content is actually an image using mimetype.
  [thomasdesvenain]


src/plone/namedfile/utils/__init__.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
.. $Id$
"""

from __future__ import division
from __future__ import print_function
from __future__ import absolute_import

import os
import six
import struct
import mimetypes
from io import BytesIO

import piexif

from PIL import Image

from plone.namedfile.utils.png_utils import process_png
from plone.namedfile.utils.jpeg_utils import process_jpeg
from plone.namedfile.utils.tiff_utils import process_tiff

logger = __import__('logging').getLogger(__name__)


def safe_basename(filename):
    """
    Get the basename of the given filename, regardless of which platform
    (Windows or Unix) it originated from.
    """
    fslice = max(
        filename.rfind('/'),
        filename.rfind('\\'),
        filename.rfind(':'),
    ) + 1
    return filename[fslice:]


def get_contenttype(source=None,
                    filename=None,
                    default='application/octet-stream'):
    """
    Get the MIME content type of the given file and/or filename.
    """
    file_type = getattr(source, 'contentType', None)
    if file_type:
        return file_type
    filename = getattr(source, 'filename', filename)
    if filename:
        return mimetypes.guess_type(filename, strict=True)[0] or 'application/octet-stream'
    return default


def bytes_(s, encoding='utf-8', errors='strict'):
    """
    If ``s`` is an instance of ``text_type``, return
    ``s.encode(encoding, errors)``, otherwise return ``s``
    """
    if not isinstance(s, bytes):
        if hasattr(s, 'encode'):
            s = s.encode(encoding, errors)
        elif hasattr(s, '__bytes__'):
            s = s.__bytes__()
    return s


def ensure_data(image):
    data = None
    if getattr(image, 'read', None):
        data = image.read()
        image.seek(0)
    else:
        data = image
    return bytes_(data)
_ensure_data = ensure_data


def getImageInfo(data):
    content_type = None
    width, height = -1, -1
    data = ensure_data(data)
    size = len(data)
    # handle GIFs
    if size >= 10 and data[:6] in (b'GIF87a', b'GIF89a'):
        content_type = 'image/gif'
        w, h = struct.unpack(b'<HH', data[6:10])
        width = int(w)
        height = int(h)
    # handle PNG
    elif data[:8] == b'\211PNG\r\n\032\n':
        content_type, width, height = process_png(data)
    # handle JPEGs
    elif data[:2] == b'\377\330':
        content_type, width, height = process_jpeg(data)
    # handle BMPs
    elif size >= 30 and data.startswith(b'BM'):
        kind = struct.unpack(b'<H', data[14:16])[0]
        if kind == 40:  # Windows 3.x bitmap
            content_type = 'image/x-ms-bmp'
            width, height = struct.unpack(b'<LL', data[18:26])
    elif size >= 8 and data[:4] in (b"II\052\000", b"MM\000\052"):
        content_type, width, height = process_tiff(data)
    # Use PIL / Pillow to determ Image Information
    elif data:
        try:
            img = Image.open(BytesIO(data))
            width, height = img.size
            content_type = img.format or ''
            if content_type.lower() == 'tiff':
                content_type = 'image/tiff'
        except Exception as e:
            logger.exception(e)
    # return
    logger.debug('Image Info (Type: %s, Width: %s, Height: %s)',
                 content_type, width, height)
    # Type must be str
    content_type = str(content_type) if content_type else ''
    return content_type, width, height


def get_exif(image):
    exif_data = None
    image_data = ensure_data(image)
    content_type, width, height = getImageInfo(image_data)
    if content_type in ('image/jpeg', 'image/tiff'):
        # Only this two Image Types could have Exif informations
        # see http://www.cipa.jp/std/documents/e/DC-008-2012_E.pdf
        try:
            exif_data = piexif.load(image_data)
        except Exception as e:
            # TODO: determine wich error really happens
            # Should happen if data is to short --> first_bytes
            logger.warn(e)
            exif_data = exif_data = {
                '0th': {
                    piexif.ImageIFD.XResolution: (width, 1),
                    piexif.ImageIFD.YResolution: (height, 1),
                }
            }
    return exif_data


def load_exif(img):
    return piexif.load(img.info['exif'])


def img_exif_data(img):
    width, height = img.size
    exif_data = {
        '0th': {
            piexif.ImageIFD.XResolution: (width, 1),
            piexif.ImageIFD.YResolution: (height, 1),
        }
    }
    return exif_data


def rotate_image(image_data, method=None):
    """
    Rotate Image if it has Exif Orientation Informations other than 1.

    Do not use PIL.Image.rotate function as this did not transpose the image,
    rotate keeps the image width and height and rotates the image around a
    central point. PIL.Image.transpose also changes Image Orientation.
    """
    orientation = 1  # if not set assume correct orrinetation --> 1
    data = _ensure_data(image_data)
    img = Image.open(BytesIO(data))

    exif_data = None
    if 'exif' in img.info:
        try:
            exif_data = load_exif(img)
        except (ValueError):
            logger.warn('Exif information currupt')

        if exif_data and piexif.ImageIFD.Orientation in exif_data['0th']:
            orientation = exif_data['0th'][piexif.ImageIFD.Orientation]

        if      exif_data \
            and (not exif_data['0th'].get(piexif.ImageIFD.XResolution) or
                 not exif_data['0th'].get(piexif.ImageIFD.YResolution)):
            exif_data['0th'][piexif.ImageIFD.XResolution] = (img.width, 1)
            exif_data['0th'][piexif.ImageIFD.YResolution] = (img.height, 1)
    if exif_data is None:
        exif_data = img_exif_data(img)

    if method is not None:
        orientation = method

    logger.debug('Rotate image with input orientation: %s', orientation)

    fmt = img.format
    if orientation == 1:  # not transform necessary
        # img = img
        pass
    elif orientation == 2:
        img = img.transpose(Image.FLIP_LEFT_RIGHT)
    elif orientation == 3:
        img = img.transpose(Image.ROTATE_180)
    elif orientation == 4:
        img = img.transpose(Image.ROTATE_180).transpose(Image.FLIP_LEFT_RIGHT)
    elif orientation == 5:
        img = img.transpose(Image.ROTATE_270).transpose(Image.FLIP_LEFT_RIGHT)
    elif orientation == 6:
        img = img.transpose(Image.ROTATE_270)
    elif orientation == 7:
        img = img.transpose(Image.ROTATE_90).transpose(Image.FLIP_LEFT_RIGHT)
    elif orientation == 8:
        img = img.transpose(Image.ROTATE_90)

    if orientation in [5, 6, 7, 8]:
        if      exif_data['0th'][piexif.ImageIFD.XResolution] \
            and exif_data['0th'][piexif.ImageIFD.YResolution]:
            exif_data['0th'][piexif.ImageIFD.XResolution], \
            exif_data['0th'][piexif.ImageIFD.YResolution] = \
                exif_data['0th'][piexif.ImageIFD.YResolution], \
                exif_data['0th'][piexif.ImageIFD.XResolution]
        else:
            exif_data['0th'][piexif.ImageIFD.XResolution], \
            exif_data['0th'][piexif.ImageIFD.YResolution] = \
                (img.width, 1), (img.height, 1)

    # set orientation to normal
    exif_data['0th'][piexif.ImageIFD.Orientation] = 1
    try:
        exif_bytes = piexif.dump(exif_data)
    except Exception as e:
        logger.warn(e)
        del exif_data['Exif'][piexif.ExifIFD.SceneType]
        # This Element piexif.ExifIFD.SceneType cause error on dump
        exif_bytes = piexif.dump(exif_data)
    output_image_data = BytesIO()
    img.save(output_image_data, format=fmt, exif=exif_bytes)
    width, height = img.size
    return output_image_data.getvalue(), width, height, exif_data
